src/server/controllers/changePassword.ts

```diff
@@ -0,0 +1,131 @@
+import { Request } from 'express';
+import { Routes } from '@/shared/model/Routes';
+import deepmerge from 'deepmerge';
+import { handleAsyncErrors } from '@/server/lib/expressWrappers';
+import { getBrowserNameFromUserAgent } from '@/server/lib/getBrowserName';
+import { logger } from '@/server/lib/logger';
+import { renderer } from '@/server/lib/renderer';
+import { ResponseWithRequestState } from '@/server/models/Express';
+import {
+  validate as validateToken,
+  change as changePassword,
+} from '@/server/lib/idapi/changePassword';
+import { FieldError } from '@/shared/model/ClientState';
+import { ChangePasswordErrors } from '@/shared/model/Errors';
+import { setIDAPICookies } from '@/server/lib/cookies';
+import { trackMetric } from '@/server/lib/trackMetric';
+import { Metrics } from '@/server/models/Metrics';
+
+const validatePasswordField = (password: string): Array<FieldError> => {
```

Comment:
do we also have any disallowed characters in our passwords?

Reply:
The short answer is no.
The long answer is all the validation is done in IDAPI for passwords on submit, however IDAPI only checks the length too. As long as we detect the password hasn't been breached, and is standard unicode we allow it. The checks we do clientside are really only sanity checks.

Reply:
nice one 👍